From f70722a84ee2b929f9a4cc1eaa24f0be696a869f Mon Sep 17 00:00:00 2001
From: Victor Costan <pwnall@chromium.org>
Date: Fri, 12 May 2017 14:26:23 -0700
Subject: [PATCH 12/12] Tag FTS3 cursor pointers in fts3ColumnMethod().

FTS3 tables have a magic column, whose name matches the table name, which
is used to obtain a pointer to the cursor for the current operation.
fts3ColumnMethod() now tags its result when used with the magic column,
and fts3FunctionArg() checks the tag and returns an error when asked to
access the magic column on behalf of the user.
---
 third_party/sqlite/src/ext/fts3/fts3.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/third_party/sqlite/src/ext/fts3/fts3.c b/third_party/sqlite/src/ext/fts3/fts3.c
index 2a24b78f3783..827769881616 100644
--- a/third_party/sqlite/src/ext/fts3/fts3.c
+++ b/third_party/sqlite/src/ext/fts3/fts3.c
@@ -3349,6 +3349,7 @@ static int fts3ColumnMethod(
     /* The extra column whose name is the same as the table.
     ** Return a blob which is a pointer to the cursor.  */
     sqlite3_result_blob(pCtx, &pCsr, sizeof(pCsr), SQLITE_TRANSIENT);
+    sqlite3_result_subtype(pCtx, '3');
   }else if( iCol==p->nColumn+2 && pCsr->pExpr ){
     sqlite3_result_int64(pCtx, pCsr->iLangid);
   }else{
@@ -3562,6 +3563,7 @@ static int fts3FunctionArg(
 ){
   Fts3Cursor *pRet;
   if( sqlite3_value_type(pVal)!=SQLITE_BLOB
+   || sqlite3_value_subtype(pVal)!='3'
    || sqlite3_value_bytes(pVal)!=sizeof(Fts3Cursor *)
   ){
     char *zErr = sqlite3_mprintf("illegal first argument to %s", zFunc);
--
2.11.0
